### PR TITLE
Hardening/orion curl rewrite

### PR DESCRIPTION
**Pull request by kzangeli**
Complete rewrite of the function 'orionCurl'.
Some of the changes:
  - the payload is now always sent to a file and -d @<path> is used for curl (internal change only) - /tmp/orionFuncTestPayload (in case no file was used by the caller)
  - it is now possible to read/understand the function :-)
  - ORION_FT_DEBUG=1 turns of debugging (output to /tmp/orionFuncTestDebug.log)
  - New parameters: -H and --header (to add headers to curl)

### CODE REVIEW

File: test/functionalTest/cases/672_forbidden_chars_in_url/all_forbidden_chars_in_uri.test [hunk 1]
```diff
@@ -58,7 +58,7 @@ echo
 
 echo '03. Test the forbidden character "'
 echo "=================================="
-orionCurl --url '/test/with/forbidden/character/"' --json
+orionCurl --url '/test/with/forbidden/character/\"' --json
```
Comment (kzangeli) [inline test/functionalTest/cases/672_forbidden_chars_in_url/all_forbidden_chars_in_uri.test]: I had to escape the " as "" is used in the call to curl.

Just informative, NTC

File: src/lib/rest/rest.cpp [hunk 2]
```diff
@@ -518,6 +518,18 @@ static char* removeTrailingSlash(std::string path)
 */
 int servicePathSplit(ConnectionInfo* ciP)
 {
+  //
+  // Special case: empty service-path 
+  //
+  if ((ciP->httpHeaders.servicePathReceived == true) && (ciP->servicePath == ""))
```
Comment (fgalan) [inline src/lib/rest/rest.cpp]: This is related with the last comments in https://github.com/telefonicaid/fiware-orion/issues/843?

File: test/functionalTest/cases/000_bad_requests/pagination_error_in_uri_params.test [hunk 3]
```diff
@@ -37,7 +37,7 @@ echo
 echo
 
 echo "+++++ 2. bad characters in URI param 'offset'"
-url="/v1/queryContext?limit=0&offset=qwe"
+url="/v1/queryContext?offset=qwe"
```
Comment (fgalan) [inline test/functionalTest/cases/000_bad_requests/pagination_error_in_uri_params.test]: Is this related with the orionCurl refactoring or only cleanup?

File: test/functionalTest/harnessFunctions.sh [hunk 4]
```diff
@@ -595,22 +625,200 @@ function dbInsertEntity()
 #   --host        <host>           (default: localhost)
 #   --port        <port>           (default: $CB_PORT)
 #   --url         <URL>            (default: empty string)
+#   --urlParams   <params>         (URI parameters 'in' URL-string)
```
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: You have defined this new parameter for orionCurl but it seems that there isn't any .test using it?

Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: I have realized that it is not a  new parameter (it was in the previous version of the function), but no .test is using, anyway...

File: test/functionalTest/harnessFunctions.sh [hunk 5]
```diff
@@ -595,22 +625,200 @@ function dbInsertEntity()
...
 #   --payload     <payload>        (default: NO PAYLOAD. Possible values: [filename | "${string}"])
 #   --in          (input payload)  (default: xml => application/xml, If 'json': application/json)
 #   --out         (output payload  (default: xml => application/xml, If 'json': application/json)
 #   --json        (in/out JSON)    (if --in/out is used AFTER --json, it overrides) 
 #   --httpTenant  <tenant>         (tenant in HTTP header)
+#   --tenant      <tenant>         (tenant in HTTP header)
```
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: Is this the new parameter actually used?

File: src/lib/rest/rest.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Might be, but that was not my intention.
I bumped into a few errors during the troubleshooting of the new implementation of orionCurl.
The changes in this PR apart from that function itself are to remedy those errors.

File: test/functionalTest/cases/000_bad_requests/pagination_error_in_uri_params.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/000_bad_requests/pagination_error_in_uri_params.test]: In this functest the idea is to get a complaint about offset 'awe'.
If I leave the 'offset=0', then it complains about offset instead.
I don't know why this has changed, but the easiest way to get it to complain as 'wanted', was to remove limit=0.
Why didn't the broker complain about limit=0 before?  Beats me ...

File: test/functionalTest/harnessFunctions.sh (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/harnessFunctions.sh]: Yeah, we normally just add '?a=1&b=2' to the URL.
Both ways can be used ...

File: test/functionalTest/harnessFunctions.sh (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/harnessFunctions.sh]: I would like to migrate from --httpTenant to --tenant.
--httpTenant is an old name, from when we had both --httpTenant and --urlTenant.
It's name should have changed when --urlTenant was removed.
Now orionCurl is prepared for this change.
I need to create an issue about this ...

Comment (kzangeli) [inline test/functionalTest/harnessFunctions.sh]: Issue #938

File: test/functionalTest/harnessFunctions.sh [hunk 6]
```diff
@@ -595,22 +625,200 @@ function dbInsertEntity()
...
 #   --servicePath <path>           (Service Path in HTTP header)
-#   --urlParams   <params>         (URI parameters 'in' URL-string)
 #   --xauthToken  <token>          (X-Auth token value)
 #   --origin      <origin>         (Origin in HTTP header)
-#   --verbose                      (verbose output)
 #   --noPayloadCheck               (don't check the payload)
+#   --header      <HTTP header>    (more headers)
+#   -H            <HTTP header>    (more headers)
```
Comment (kzangeli) [inline test/functionalTest/harnessFunctions.sh]: These two options are new: -H and --header

File: src/lib/rest/rest.cpp (hunk 2 above)
Comment (fgalan) [inline src/lib/rest/rest.cpp]: If we have fixed something in the code, I think it should mentioned in CHANGES_NEXT_RELEASE.

Comment (fgalan) [inline src/lib/rest/rest.cpp]: Discussed by skype: let's try to figure out if removing this block of code is fixing something and, in that case, include a .test harness to cover the case and an issue to be added to CHANGES_NEXT_RELEASE.

Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Commented with FIXME in b507e8f

File: test/functionalTest/cases/000_bad_requests/pagination_error_in_uri_params.test (hunk 3 above)
Comment (fgalan) [inline test/functionalTest/cases/000_bad_requests/pagination_error_in_uri_params.test]: NTC

File: test/functionalTest/harnessFunctions.sh (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: Great! More flexibility...

(In fact, `--origin` could have been implemented using `-H`, but let's keep it as it is now).

NTC

File: test/functionalTest/testHarness.sh [hunk 7]
```diff
@@ -41,6 +41,40 @@ cd - > /dev/null 2>&1
...
+  _debug='on'
+fi
+
+
+
+# ------------------------------------------------------------------------------
+#
+# dMsg2 - debug output
+#
+function dMsg2()
```
Comment (fgalan) [inline test/functionalTest/testHarness.sh]: dMsg2? Should use a more descriptive name...

File: test/functionalTest/harnessFunctions.sh (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: NTC

File: test/functionalTest/testHarness.sh (hunk 7 above)
Comment (kzangeli) [inline test/functionalTest/testHarness.sh]: Fixed in 5fef94f

**Comment by fgalan** [comment]
LGTM

**Pull request merged**